host/monitor.go

```diff
@@ -0,0 +1,213 @@
+package main
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	"github.com/flynn/flynn/discoverd/client"
+	"github.com/flynn/flynn/host/fixer"
+	"github.com/flynn/flynn/pkg/cluster"
+)
+
+const (
+	checkInterval = 10 * time.Second
+)
+
+var monitorLogger = log15.New("component", "cluster-monitor")
+
+type MonitorMetadata struct {
+	Enabled bool `json:"enabled,omitempty"`
+	Hosts   int  `json:"hosts,omitempty"`
+}
+
+type Monitor struct {
+	addr       string
+	dm         *DiscoverdManager
+	discoverd  *discoverdWrapper
+	discClient *discoverd.Client
+	monitorSvc discoverd.Service
+	isLeader   bool
+	c          *cluster.Client
+	hostCount  int
+	deadline   time.Time
+	shutdownCh chan struct{}
+}
+
+func NewMonitor(dm *DiscoverdManager, addr string) *Monitor {
+	return &Monitor{
+		dm:         dm,
+		discoverd:  nil,
+		addr:       addr,
+		shutdownCh: make(chan struct{}),
+	}
+}
+
+func (m *Monitor) waitDiscoverd() {
+	for {
+		if m.dm.localConnected() {
+			m.discClient = discoverd.NewClient()
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (m *Monitor) waitRaftLeader() {
+	for {
+		_, err := m.discClient.RaftLeader()
+		if err == nil {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (m *Monitor) waitEnabled() {
+	for {
+		monitorMeta, err := m.monitorSvc.GetMeta()
+		if err != nil {
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		var decodedMeta MonitorMetadata
+		if err := json.Unmarshal(monitorMeta.Data, &decodedMeta); err != nil {
+			continue
```

Comment:
This will result in a tight loop fetching the metadata if the JSON is unparseable, log and add a delay?